docs/determinism.md

```diff
@@ -0,0 +1,13 @@
+# Determinism (GPU)
+
+## Compilation
+
+XLA is generally developed such that the compilation process is deterministic by choosing appropriate algorithms and data structures.
+
+One important exception is autotuning, in particular of GEMM fusions: due to fluctuations in measurements different kernels can be picked as the fastest ones in different compilation runs. [Persisted autotuning](./persisted_autotuning) can be used as a workaround.
+
+Another small exception is string serialization of protobufs with hash maps like fusion backend configs: these are not printed in deterministic order. This leads to variations in HLO IR dumps and can cause bugs if backend configs are used to compare HLO instructions. See the [helper methods](https://github.com/openxla/xla/blob/a02433bdc5e6dbe68446d6c95c4959b5e1715cf9/xla/service/gpu/ir_emission_utils.h#L242-L265) added to work around this problem.
+
+## Execution
+
+Programs compiled by XLA can be non-deterministic on operations like scatter, select-and-scatter, GEMMs, convolutions, multi-headed attention. Flag `--xla_gpu_exclude_nondeterministic_ops` switches these operations to deterministic and potentially slower implementations and makes compilation fail on select-and-scatter which does not have a deterministic implementaiton.
```

Comment:
Maybe mention how that flag can be combined with the persisted autotuning?

Reply:
As usual - by chaining flags. I don't see why does it require separate documentation.

Reply:
This can be done easily by just telling that it can be combined with persistent autotuner cache.
This is a special workflow. So it warent something being written. A simple paragraph like:

You want determinsitic computation, but the minimal slowdown? You can recover some of the speed lost by filling the autotuner cache first (run without the deterministic flag), then re-run with the cache full and with the deterministic flag on. XLA will use the autotuner cache in that setup, and it something is missing, it will crash.

I'm sure you can write it better then what I wrote above.

Reply:
Closing. The current version address this issue already.